### PR TITLE
Also abort ongoing file restores when snapshot restore is aborted

### PR DESCRIPTION
**Pull request by tlrx**
Today when a snapshot restore is aborted (for example when the index is explicitly deleted) while the restoration of the files from the repository has already started the file restores are not interrupted. It means that Elasticsearch will continue to read the files from the repository and will continue to write them to disk until all files are restored; the store will then be closed and files will be deleted from disk at some point but this can take a while. This will also take some slots in the SNAPSHOT thread pool too. The Recovery API won't show any files actively being recovered, the only notable indicator would be the active threads in the SNAPSHOT thread pool.

This pull request adds a check before reading a file to restore and before writing bytes on disk so that a closing store can be detected more quickly and the file recovery process aborted. This way the file restores just stops and for most of the repository implementations it means that no more bytes are read (see #62370 for S3), finishing threads in the SNAPSHOT thread pool more quickly too.

### CODE REVIEW

**Comment by tlrx** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows ([failure](https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request+packaging-tests-windows-gcp-sample/8272/))

File: server/src/main/java/org/elasticsearch/index/store/Store.java [hunk 1]
```diff
@@ -403,14 +403,21 @@ public final void decRef() {
...
             decRef();
             logger.debug("store reference count on close: {}", refCounter.refCount());
         }
     }
 
+    /**
+     * @return true if the {@link Store#close()} method has been called indicating that the current store is being closed but potentially
+     * not yet fully closed and released. You might prefer to use {@link Store#ensureOpen()} instead.
+     */
+    public boolean isClosing() {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/index/store/Store.java:417]: I'd reword this a little:

```java
/**
 * @return true if the {@link Store#close()} method has been called. This indicates that the current 
 * store is either closed or being closed waiting for all references to it to be released. 
 * You might prefer to use {@link Store#ensureOpen()} instead.
 */
```

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/AbortedRestoreIT.java [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+
+        final String indexName = "test-abort-restore";
+        final int numPrimaries = randomIntBetween(1, 3);
+        createIndex(indexName, indexSettingsNoReplicas(numPrimaries).build());
+        indexRandomDocs(indexName, scaledRandomIntBetween(10, 1_000));
+        ensureGreen();
+        forceMerge();
+
+        final String repositoryName = "repository";
+        createRepository(repositoryName, BlockingDataFileReadsRepository.TYPE, Settings.builder().put("location", randomRepoPath()));
```
Comment (original-brownbear) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/AbortedRestoreIT.java:90]: See below comment: I think you can just use `createRepository(repositoryName, BlockingDataFileReadsRepository.TYPE)` if you check the store on every chunk read?

**Comment by tlrx** [comment]
@elasticmachine update branch

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/AbortedRestoreIT.java [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+     * is set to true. It also keep track of the number of bytes read from {@link InputStream} it opens.
+     */
+    public static class BlockingDataFileReadsRepository extends FsRepository {
+
+        static final String TYPE = "block_on_data_file_reads";
+
+        private final List<BlockingInputStream> streams = Collections.synchronizedList(new ArrayList<>());
+        private volatile boolean blockStreams = false;
+
+        public BlockingDataFileReadsRepository(
```
Comment (original-brownbear) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/AbortedRestoreIT.java:186]: Maybe just override `org.elasticsearch.snapshots.mockstore.MockRepository#blockOnDataFiles` and add a method to verify no more interactions (could just add a flag that throws an `AssertionError` past a certain point)? Then you can just wait for the snapshot threads to all block, unblock and wait for the snapshot pool to be empty. Much less new test code to maintain and tests the same thing pretty much?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 4]
```diff
@@ -2101,9 +2107,10 @@ protected InputStream openSlice(int slice) throws IOException {
                                     return container.readBlob(fileInfo.partName(slice));
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:2107]: Maybe put `ensureNotClosing` here as well? Then you can run your test with random chunk_size as well I think?

**Review by original-brownbear — COMMENTED** [review]
Nice find @tlrx, I have some suggestions on the tests , let me know what you think :)

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/AbortedRestoreIT.java (hunk 3 above)
Comment (tlrx) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/AbortedRestoreIT.java:186]: I agree it would be less code to maintain, but `MockRepository` blocks before reading any bytes and I wanted something that allows to read few bytes then block. But now I'm thinking to it again it does not make more difference as `InputStream` opened by FS blob containers do not have any abort method. I'm going to take a look again at using `MockRepository`.

File: server/src/main/java/org/elasticsearch/index/store/Store.java (hunk 1 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/index/store/Store.java:417]: It's better, thanks!

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 4 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:2107]: Sure, makes sense. Not sure why I missed it

**Comment by tlrx** [comment]
Thanks @original-brownbear - I've applied your suggestions, let me know what you think please

**Review by original-brownbear — COMMENTED** [review]
Two open questions on the tests, otherwise looks very nice thanks!

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/AbortedRestoreIT.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            Environment environment,
+            NamedXContentRegistry namedXContentRegistry,
+            ClusterService clusterService,
+            RecoverySettings recoverySettings
+        ) {
+            super(metadata, environment, namedXContentRegistry, clusterService, recoverySettings);
+        }
+
+        @Override
+        protected int bufferSize() {
```
[the remark this answers is not in the document]
Comment (tlrx) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/AbortedRestoreIT.java]: Agreed, I reverted this change

**Comment by tlrx** [comment]
Thanks for insisting on details Armin. It should be good now :)

**Review by original-brownbear — APPROVED** [review]
LGTM thanks for fixing all those annoying details!

**Comment by tlrx** [comment]
Thanks Armin!
